app/assets/javascripts/angular/directives/graph_chart.js

```diff
@@ -11,6 +12,147 @@ angular.module("Prometheus.directives").directive('graphChart', ["$location", "W
+                break;
+              }
+
+              if (data[i + 1].x <= domainX) { i++ } else { i-- }
+            }
+
+            if (dataIndex < 0) dataIndex = 0;
+            var value = data[dataIndex];
+
+            var distance = Math.sqrt(
+              Math.pow(Math.abs(graph.x(value.x) - eventX), 2) +
+              Math.pow(Math.abs(graph.y(value.y + value.y0) - eventY), 2)
+            );
+
+            var point = {
+              series: series,
+              value: value,
+              distance: distance,
+              order: j,
+              name: series.name
+            };
+
+            if (!nearestPoint || distance < nearestPoint.distance) {
+              nearestPoint = point;
+            }
+
+            points.push(point);
+          }, this );
+
+          if (nearestPoint) {
+            return nearestPoint.value.x;
+          }
+        }
+
+        function onMousedown() {
+          rectangle = svgContainer.append("rect")
+                      .style("opacity", 0.5)
+                      .attr("y", 0)
+                      .attr("height", "100%");
+
+          drag.target = d3.event.target;
+          drag.startDt = pointAsDate(d3.event);
+          drag.y = d3.event.offsetY;
+          drag.startPX = d3.event.offsetX;
+          d3.event.preventDefault ? d3.event.preventDefault() : d3.event.returnValue = false;
+          d3.select(document).on("mousemove", onMousemove);
+          d3.select(document).on("mouseup", onMouseup);
+          d3.select(document).on("keyup", function() {
+            if (d3.event.keyCode === 27) {
+              reset();
+            }
+          });
+          d3.select(document).on("touchmove", onMousemove);
+          d3.select(document).on("touchend, touchcancel", onMouseup); // i think i can listen to both events like this..
+          // d3.select(document).on("touchcancel", onMouseup);
+        }
+
+        function onMouseup(datum, index) {
+          drag.stopDt = pointAsDate(d3.event);
+          var windowAfterDrag = [
+            drag.startDt,
+            drag.stopDt
+          ].sort();
+
+          graph.window.xMin = windowAfterDrag[0];
+          graph.window.xMax = windowAfterDrag[1];
+          graph.update();
+
+          reset();
+        }
+
+        function reset() {
+          d3.select(document).on("mousemove", null);
+          d3.select(document).on("mouseup", null);
+          d3.select(document).on("touchmove", null);
+          d3.select(document).on("touchend, touchcancel", null); // i think i can listen to both events like this..
+          drag = {};
+          rectangle.remove()
+        }
+
+        function onMousemove() {
+          //Draw the rectangle
+          var limits = [drag.startPX, d3.event.offsetX].sort();
+          rectangle
+          .attr("fill", "steelblue")
+          .attr("x", limits[0])
+          .attr("width", limits[1]-limits[0]);
```

Comment:
ah yeah, that's how my vim does it, and i was in "free-hack" mode, so i was just going with it